sys/kern/interrupt.c

```diff
@@ -127,60 +141,103 @@ void intr_root_handler(ctx_t *ctx) {
     on_user_exc_leave();
 }
 
-/* interrupt handlers delegated to be called in the interrupt thread */
-static ih_list_t delegated = TAILQ_HEAD_INITIALIZER(delegated);
+static void intr_thread_create(intr_event_t *ie) {
+  intr_thread_t *it = kmalloc(M_INTR, sizeof(intr_thread_t), M_ZERO);
+
+  assert(it);
+
+  ie->ie_ithread = it;
+
+  it->it_thread = thread_create(ie->ie_name, intr_thread, it, prio_ithread(0));
+
+  assert(it->it_thread != NULL);
+
+  sched_add(it->it_thread);
+}
 
 void intr_event_run_handlers(intr_event_t *ie) {
   intr_handler_t *ih, *next;
   intr_filter_t status = IF_STRAY;
 
   assert(ie != NULL);
 
+  /* Do we wake up an ithread */
+  bool it_wakeup = false;
+
   TAILQ_FOREACH_SAFE (ih, &ie->ie_handlers, ih_link, next) {
     status = ih->ih_filter(ih->ih_argument);
     if (status == IF_FILTERED)
-      return;
+      continue;
     if (status == IF_DELEGATE) {
       assert(ih->ih_service);
 
       if (ie->ie_disable)
         ie->ie_disable(ie);
 
-      TAILQ_REMOVE(&ie->ie_handlers, ih, ih_link);
-      TAILQ_INSERT_TAIL(&delegated, ih, ih_link);
-      sleepq_signal(&delegated);
-      return;
+      ih->ih_flags = IH_DELEGATE;
+
+      it_wakeup = true;
     }
   }
 
+  if (it_wakeup) {
+    if (ie->ie_ithread == NULL)
+      intr_thread_create(ie);
+
+    sleepq_signal(ie->ie_ithread);
+  }
+
   klog("Spurious %s interrupt!", ie->ie_name);
```

Comment:
You will always reach this line, hence every correct interrupt request will yield the "Spurious interrupt!" message. To prevent this, you can add a `bool stray = true;` value and clear it if a handler is filtered or delegated. Then at this point do `if (stray) klog(...)`.